### PR TITLE
ytt: init at 0.30.0

### PR DESCRIPTION
**Pull request by brhoades**
https://github.com/vmware-tanzu/carvel-ytt

###### Motivation for this change
ytt is not already available in nixpkgs. Tests ran by checkPhase fail due to YAML marshaling issues but failures are unrelated to the result binary.

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- N/A  Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by wilsonehusin — APPROVED** [review]
LGTM

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -1252,6 +1252,16 @@
     githubId = 3043718;
     name = "Brett Lyons";
   };
+  brodes = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:1255]: Please put this into a separate commit with the message ``maintainers: add brodes``.

File: pkgs/development/tools/ytt/default.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{ pkgs, lib, ... }:
+with pkgs; with lib; buildGoModule rec {
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:2]:
```suggestion
{  lib }:
buildGoModule rec {
```

Don't use with for the entire scope. Also don't pass in pkgs. Triple dots are only needed if you can add any number of extra packages which is not needed here.

File: pkgs/development/tools/ytt/default.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  version = "0.30.0";
+
+  goPackagePath = "github.com/vmware-tanzu/carvel-ytt";
+  src = fetchFromGitHub {
+    sha256 = "0v9wp15aj4r7wif8i897zwj3c6bg41b95kk7vi3a3bzin814qn6l";
+    rev = "v${version}";
+    repo = "carvel-ytt";
+    owner = "vmware-tanzu";
+  };
+  vendorSha256 = null;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:13]:
```suggestion
  src = fetchFromGitHub {
    owner = "vmware-tanzu";
    repo = "carvel-ytt";
    rev = "v${version}";
    sha256 = "0v9wp15aj4r7wif8i897zwj3c6bg41b95kk7vi3a3bzin814qn6l";
  };

  goPackagePath = "github.com/vmware-tanzu/carvel-ytt";

  vendorSha256 = null;
```

File: pkgs/development/tools/ytt/default.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  # on v0.30.0 YAML marshalling tests fail without consequence for ytt itself
+  checkPhase = "";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:16]:
```suggestion
  doCheck = false;
```

File: pkgs/development/tools/ytt/default.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  preFixup = ''
+    rm -f $out/bin/ytt-lambda-website
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:18]:
```suggestion
    rm $out/bin/ytt-lambda-website
```

Otherwise we won't notice when something changes.

File: pkgs/development/tools/ytt/default.nix [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  '';
+
+  meta = {
+    description = ''
+      YAML templating tool that allows configuration of complex software via reusable templates with user-provided values
+    '';
+
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:25]:
```suggestion
    description = "YAML templating tool that allows configuration of complex software via reusable templates with user-provided values";
```

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -28655,4 +28655,5 @@ in
 
   psftools = callPackage ../os-specific/linux/psftools {};
 
+  ytt = callPackage ../development/tools/ytt {};
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:28658]: Please put this under the tools headline and not at the end of everything.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/ytt/default.nix (hunk 2 above)
Comment (brhoades) [reply pkgs/development/tools/ytt/default.nix:2]: Thanks for the suggestions. I'll definitely remove the generous use of with, but unless I'm missing something, I'll need to keep `pkgs` as it's required for `buildGoModule`.

Comment (brhoades) [reply pkgs/development/tools/ytt/default.nix:2]: I've figured this out: I can just put required functions in this function's set and it'll work out.

File: pkgs/top-level/all-packages.nix (hunk 7 above)
Comment (brhoades) [reply pkgs/top-level/all-packages.nix]: Didn't see the sections on the file, but I now see the file-based sorting. thanks.

File: pkgs/development/tools/ytt/default.nix [hunk 8]
```diff
@@ -0,0 +1,30 @@
+{ lib, buildGoModule, fetchFromGitHub }:
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:1]: Please update the PR title and commit message to ``ytt: init at 0.30.0``.

File: pkgs/development/tools/ytt/default.nix [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+  # on v0.30.0 YAML marshalling tests fail without consequence for ytt itself
+  doCheck = false;
+  preFixup = ''
+    rm $out/bin/ytt-lambda-website
+  '';
+
+  meta = with lib; {
+    description = "YAML templating tool that allows configuration of complex software via reusable templates with user-provided values";
+
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:25]:
```suggestion
```

File: pkgs/development/tools/ytt/default.nix [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    repo = "carvel-ytt";
+    rev = "v${version}";
+    sha256 = "0v9wp15aj4r7wif8i897zwj3c6bg41b95kk7vi3a3bzin814qn6l";
+  };
+
+  goPackagePath = "github.com/vmware-tanzu/carvel-ytt";
+
+  vendorSha256 = null;
+
+  # on v0.30.0 YAML marshalling tests fail without consequence for ytt itself
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:17]: It is easily possible to disable this one test?

File: pkgs/top-level/all-packages.nix [hunk 11]
```diff
@@ -28654,5 +28656,4 @@ in
   cagebreak = callPackage ../applications/window-managers/cagebreak/default.nix {};
 
   psftools = callPackage ../os-specific/linux/psftools {};
-
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:28657]: Please revert this change.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 105071` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ytt</li>
  </ul>
</details>

File: pkgs/development/tools/ytt/default.nix (hunk 10 above)
Comment (brhoades) [reply pkgs/development/tools/ytt/default.nix]: Yep, in fact their [developer documentation](https://github.com/vmware-tanzu/carvel-ytt/blob/develop/docs/dev.md) refers to a script which does just that. Since the script has other dependencies, I've pulled out the important line into a custom checkPhase.

File: pkgs/development/tools/ytt/default.nix [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+  vendorSha256 = null;
+
+  # extracted from /hack/test-unit.sh
+  checkPhase = ''
+    go test "$(go list ./... | grep -v yaml.v2)"
+  '';
+
+  preFixup = ''
+    rm $out/bin/ytt-lambda-website
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:23]: Can we set ``subPackages = [ "cmd/something" ];`` to only build the required binary?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/ytt/default.nix (hunk 12 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/ytt/default.nix:23]: Even without this change I get:

```
running tests
cannot find package "." in:
        /private/tmp/nix-build-ytt-0.30.0.drv-0/source/cmd/ytt
github.com/k14s/ytt/cmd/ytt-lambda-website
github.com/k14s/ytt/pkg/cmd
github.com/k14s/ytt/pkg/cmd/core
github.com/k14s/ytt/pkg/cmd/template
github.com/k14s/ytt/pkg/filepos
github.com/k14s/ytt/pkg/files
github.com/k14s/ytt/pkg/orderedmap
github.com/k14s/ytt/pkg/structmeta
github.com/k14s/ytt/pkg/template
github.com/k14s/ytt/pkg/template/core
github.com/k14s/ytt/pkg/texttemplate
github.com/k14s/ytt/pkg/version
github.com/k14s/ytt/pkg/website
github.com/k14s/ytt/pkg/workspace
github.com/k14s/ytt/pkg/yamlfmt
github.com/k14s/ytt/pkg/yamlmeta
github.com/k14s/ytt/pkg/yamltemplate
github.com/k14s/ytt/pkg/yttlibrary
github.com/k14s/ytt/pkg/yttlibrary/overlay
```

File: pkgs/development/tools/ytt/default.nix [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  '';
+
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:25]:
```suggestion
```

File: pkgs/development/tools/ytt/default.nix [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    owner = "vmware-tanzu";
+    repo = "carvel-ytt";
+    rev = "v${version}";
+    sha256 = "0v9wp15aj4r7wif8i897zwj3c6bg41b95kk7vi3a3bzin814qn6l";
+  };
+
+  goPackagePath = "github.com/vmware-tanzu/carvel-ytt";
+
+  vendorSha256 = null;
+
```
Comment (SuperSandro2000) [inline pkgs/development/tools/ytt/default.nix:16]:
```suggestion

subPackages = [ "cmd/ytt" ];

```

Comment (brhoades) [reply pkgs/development/tools/ytt/default.nix:16]: I'll do this, but there are no tests ran since they're tucked away in other subpackages. It does make for a smaller file overall.

File: pkgs/development/tools/ytt/default.nix (hunk 12 above)
Comment (brhoades) [reply pkgs/development/tools/ytt/default.nix:23]: Setting subPackages to `cmd/ytt` has the effect of only building the desired command. We can't run any tests this way, but overall should be faster.

I'm fine with doing it this way and only tried to include tests since they were erroring.

**Review by SuperSandro2000 — COMMENTED** [review]
Please change the title and commit message to not contain a v in the version number.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 105071` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ytt</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 105071` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ytt</li>
  </ul>
</details>

**Comment by brhoades** [comment]
Change the title and commit.

**Comment by SuperSandro2000** [comment]
Congratulations on your first PR.

**Comment by brhoades** [comment]
Thanks for the assistance!
